hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/compact/HoodieCompactor.java

```diff
@@ -278,4 +368,123 @@ public CompactionExecutionHelper getCompactionExecutionStrategy(HoodieCompaction
     }
   }
 
+  List<WriteStatus> compactWithFileGroupReader(HoodieReaderContext readerContext,
+                                               String instantTime,
+                                               HoodieTableMetaClient metaClient,
+                                               CompactionOperation operation,
+                                               List<HoodieLogFile> logFiles,
+                                               Schema readerSchema,
+                                               Option<InternalSchema> internalSchemaOpt,
+                                               TypedProperties props,
+                                               HoodieWriteConfig config,
+                                               TaskContextSupplier taskContextSupplier) throws IOException {
+    HoodieTimer timer = HoodieTimer.start();
+    timer.startTimer();
+    Option<HoodieBaseFile> baseFileOpt =
+        operation.getBaseFile(metaClient.getBasePath().toString(), operation.getPartitionPath());
+    FileSlice fileSlice = new FileSlice(
+        operation.getFileGroupId(),
+        operation.getBaseInstantTime(),
+        baseFileOpt.isPresent() ? baseFileOpt.get() : null,
+        logFiles);
+
+    // 1. Generate the input for fg reader.
+    boolean usePosition = config.getBooleanOrDefault(MERGE_USE_RECORD_POSITIONS);
+    HoodieFileGroupReader<T> fileGroupReader = new HoodieFileGroupReader<>(
+        readerContext,
+        metaClient.getStorage(),
+        metaClient.getBasePath().toString(),
+        instantTime,
+        fileSlice,
+        readerSchema,
+        readerSchema,
+        internalSchemaOpt,
+        metaClient,
+        props,
+        0,
+        Long.MAX_VALUE,
+        usePosition);
+
+    // 2. Get the `HoodieFileGroupReaderIterator` from the fg reader.
+    fileGroupReader.initRecordIterators();
+    HoodieFileGroupReader.HoodieFileGroupReaderIterator<T> recordIterator
+        = fileGroupReader.getClosableIterator();
+
+    // 3. Write the record using parquet writer.
+    String writeToken = FSUtils.makeWriteToken(
```

Comment:
hey @yihua : Should we use HoodieCreateHandle here. 
I am worried about adhoc code written here. what incase we miss out something which regular MergeHandle should do?

Reply:
see. we are not creating markers here. thats the reason I told I was not comfortable making this default for whole of spark. 
unless we go through one of the existing write handles, the new custom code we write might have some gaps.

Reply:
For eg, does this handle virtual keys?

Reply:
CC: @yihua